x-pack/plugins/endpoint/public/applications/endpoint/store/hosts/reducer.ts

```diff
@@ -38,30 +42,77 @@ export const hostListReducer: ImmutableReducer<HostListState, AppAction> = (
       pageSize,
       pageIndex,
       loading: false,
+      error: undefined,
+    };
+  } else if (action.type === 'serverFailedToReturnHostList') {
+    return {
+      ...state,
+      error: action.payload,
+      loading: false,
     };
   } else if (action.type === 'serverReturnedHostDetails') {
     return {
       ...state,
       details: action.payload.metadata,
+      detailsLoading: false,
+      detailsError: undefined,
     };
   } else if (action.type === 'serverFailedToReturnHostDetails') {
     return {
       ...state,
       detailsError: action.payload,
+      detailsLoading: false,
     };
-  } else if (action.type === 'userPaginatedHostList') {
-    return {
+  } else if (action.type === 'userChangedUrl') {
+    const newState: Immutable<HostState> = {
       ...state,
-      ...action.payload,
-      loading: true,
+      location: action.payload,
     };
-  } else if (action.type === 'userChangedUrl') {
+    const isCurrentlyOnListPage = isOnHostPage(newState) && !hasSelectedHost(newState);
+    const wasPreviouslyOnListPage = isOnHostPage(state) && !hasSelectedHost(state);
+    const isCurrentlyOnDetailsPage = isOnHostPage(newState) && hasSelectedHost(newState);
+    const wasPreviouslyOnDetailsPage = isOnHostPage(state) && hasSelectedHost(state);
+
+    // if on the host list page for the first time, return new location and load list
+    if (isCurrentlyOnListPage) {
+      if (!wasPreviouslyOnListPage) {
+        return {
+          ...state,
+          location: action.payload,
+          loading: true,
+          error: undefined,
+          detailsError: undefined,
+        };
+      }
+    } else if (isCurrentlyOnDetailsPage) {
+      // if previous page was the list or another host details page, load host details only
+      if (wasPreviouslyOnDetailsPage || wasPreviouslyOnListPage) {
+        return {
+          ...state,
+          location: action.payload,
+          detailsLoading: true,
+          error: undefined,
+          detailsError: undefined,
+        };
+      } else {
+        // if previous page was not host list or host details, load both list and details
+        return {
+          ...state,
+          location: action.payload,
+          loading: true,
+          detailsLoading: true,
+          error: undefined,
+          detailsError: undefined,
+        };
+      }
+    }
+    // otherwise we are not on a host list or details page
     return {
```

Comment:
You probably should reset the entire state as in:

```
return initialState();
```

Reply:
i know there's something going on to move the location out of all the individual reducers but that hasn't merged yet right? initially i did it that way so that the location would still be updated in the host reducer